### PR TITLE
xstate migration

### LINKED ISSUE #228: XState: migrate from zustand
XState is a more complex state lib, but will scale better for some of the complexities of our app. It also has some really nice dev tools to help reason about you state (visual editor etc).

This migration is worth doing sooner than later because it will a big impact on how we handle user actions, and since we want to have these fire from both buttons and command palette we'd be creating more work for ourselves by doing this later.

### LINKED ISSUE #603: disable extrude button/action for sketches that aren't closed
and in the case a user manually types it out, we should make sure that the error for that call expression makes it into the lsp.

Sample of code that doesn't work:
```typescript
const part001 = startSketchAt([5.5229, 5.25217])
  |> line([10.50433, -1.19122], %)
  |> line([8.01362, -5.48731], %)
  |> line([-1.02877, -6.76825], %)
  |> line([-11.53311, 2.81559], %)
  |> extrude(4, %)
```

error from the engine:
```
Error in response to request 4bf5a820-7a19-4e89-afe0-4070a05a435b:
bad_request: Unable to find solid2D within this object to extrude from
```

### LINKED ISSUE #654: (State fail?) While in sketch
### System
Macbook M1, Ventura 13.4.1

### Description
When pressing "Edit Sketch" on a sketch that's already created it totally sh*t's the bed.  It gets worse when you subsequently click "line" to add another line to the sketch.  
1.  When you initially click "Edit Sketch" it does not return to the sketch view. 
2.  When you click "line" and then click the model, the whole model disappears.  
3.  Recovery is now impossible and the application must restart

### Severity: High

image

### LINKED ISSUE #655: Stream blurred after trying "extrude as new" on a line
Edit by Kurt: #682 has fixed the immediate problem of getting stuck on blur, however extrude being available on sketches that isn't closed should get follow up with the xstate migration.

---

Shown by Kegan in Discord

Frank, I have a feeling this has something to do with the blur you used to have when the user drags the screen, but I might be wrong

https://github.com/KittyCAD/modeling-app/assets/114548659/7fcc68e7-d535-4668-abaf-faa5a374d1cd

### PR DESCRIPTION
**Pull request by Irev-Dev**
Resolves #228

Big refactor coming to an end.
I've already got a clean up issue here for things that I don't think should block getting this one in https://github.com/KittyCAD/modeling-app/issues/817.

State diagram here though doesn't fit into a single image super well

image

Mind you there are things not being used that should get clean up as part of #817

Walk through of the state diagram

https://github.com/KittyCAD/modeling-app/assets/29681384/d91c554c-2fe2-4577-8f1f-4defdbcadfdf

Xstate has some useful dev-tools. The one Frank and I have mostly been using for editing the diagram is the vscode extension. Please use it to do your own tour, you can have a look at the actions that are fired on events/state and then look at the implementation to see what's actually happening in terms of engine scene commands, or executing or moding code. I've also enabled a live state diagram that opens in a new tab when running dev mode (this is what I've been recording for demos). Hopefully, this isn't annoying.

Some points of context in no particular order

I think a good way of conceptualizing how we're using xstate is it maps our what's possible depending on where the user is in our app and orchestrates side-effects. The side-effects I'm talking about is both the engine in the form of scene-commands, but it includes all of the language state because I explicitly decided not to have the lang state managed by xstate since we want to port it rust/wasm soon anyway, more on that below.

Language state is NOT managed by xstate (specifically the code, ast, programMemory, errors and isExecuting bool), as long-term we want this state managed by rust/wasm anyway so it didn't make sense to put into xstate context. And to that end it's instead all been put into a class in `src/lang/KclSinglton.tsx`, my hope was that it's only one step away from being ported to rust, since it now doesn't have much coupling with react (beside callbacks) (cc @jessfraz, @adamchalmers, this class is very ugly and verbose with getters and setters for everything, setters in particular is important because it means we can fire a callback when ever some bit of state changes which is how we can ensure react is aware of changes, there might be a neater way to get this to work with rust, but so far it's only 5 bits of state, not too bad)

When react needs access to lang stuff to render it (code for the editor or ast for the ast explorer or program memory for the variables panel etc) then it uses the custom hook `useKclContext` so that it will get react to re-render as it needs to. Everywhere else, callbacks, xstate-actions etc, it uses the kclManager directly. Maybe this was a bad call, but it has less boilerplate associated with it because it's just one import and no need to call a custom hook beforehand. There shouldn't be an issue with stale closures because the class uses getters for everything so it should always have up-to-date state. (I think I at least haven't encountered any issues with stale data). If this is too confusing we can use `useKclContext` everywhere.

> xstate orchestrates our side-effects

This can be seen in Stream.tsx, as for a lot of the mouse related scene commands that we send over the unreliable channel change depending on what state we're currently in, so there's plenty of `if(state.match('...'))` in that file

@jgomez720 Please also try and break it

Resolves #655
Resolves #654
Resolves #603

### CODE REVIEW

File: src/Toolbar.tsx
Comment (Irev-Dev) [inline src/Toolbar.tsx:292]: @franknoirot Not sure if we need `state.value` or similar here now?

**Comment by jgomez720** [comment]
It works pretty smoothly, even with what I point out in this video I'd be fine with merging it. Xstate is going to be massive and I'd rather everyone start working with it now with some little bugs than later
https://github.com/KittyCAD/modeling-app/assets/114548659/eb92ff0f-ef7b-4f7d-b146-c3dc8fb232e5

**Comment by franknoirot** [comment]
Sorry I'm a little out-of-date on what I should expect for feature parity. Should I be able to edit a sketch after completing it? I tried to edit my sketch and use the Move tool and I'm pretty reliably able to break the stream doing it.

**Comment by franknoirot** [comment]
The big thing from testing is that I confirmed:
- Re-entering and exiting sketches repeatedly works
- Constraint buttons correctly disable if your selection doesn't meet the criteria
- Contraints are correctly applied
  - First setting a Length on a line and then making it Horizontal felt really good: the constant was correctly preserved!

So I don't know if any of this feedback rises to the level of blocking merge. I agree with @jgomez720 that this is such a big pivot that it will necessarily require bug fixes after the fact regardless.

Some user testing notes:
- Closing a sketch line results in weird, hard-to-describe behavior
  - Move tool no longer works
  - Vertex selection indicators no longer appear
  - Selecting lines becomes unreliable (might be an unrelated network congestion issue I was seeing though)
- Applying a Length constraint to a line doesn't clear the existing entities from the scene before drawing the new ones, resulting in a stale double-image:
  - Screenshot 2023-10-10 at 2 51 09 PM
- Move command doesn't seem to work more than once?
- I seem to be able to use the Move tool to move points that have been constrained. Code mods don't apply though which is probably a good sign; I bet I'm just seeing enhancements we want to do post-merge:

https://github.com/KittyCAD/modeling-app/assets/23481541/c67bd0c6-7a75-40e9-ade4-f35ba139d59c

Okay I'll review the code now.

**Comment by jessfraz** [comment]
are any of these also fixed by this https://github.com/KittyCAD/modeling-app/issues?q=is%3Aissue+is%3Aopen+label%3Axstate just wanna close any and all bugs we can or those can be done after initial merge

**Comment by Irev-Dev** [comment]
@franknoirot feature parity should be there unless it's very minor, so yeah I'll look at the edit sketch thing now.

> Closing a sketch line results in weird, hard-to-describe behavior

I thought this was existing behaviour but I might be wrong. After a close it re-executes and then the engine and app are kinda out of sync, so I'd consider it a known issue that's common to both.

> Applying a Length constraint to a line doesn't clear the existing entities from the scene before drawing the new ones
That's weird I suspect it's something small, but again happened before the migration

Yeah writing this next bit post the app sync, a lot of little interactions with the sketches are broken and were already broken.

**Review by franknoirot — APPROVED** [review]
Let's go! I think all my notes are non-blocking. Huge effort, and I think well worth it. Really excited about this.

File: src/lang/queryAst.ts
Comment (Irev-Dev) [reply src/lang/queryAst.ts:1]: Yeah it looks for if a pipExpression has a particular callExpression, let me rename it,  I think `isCallExpressionInPipe`, hmm maybe. `doesPipeHaveCallExp('LineTo')` is maybe better.
a1dae8f729d7e88f49fb04a88b53f8f9185243b1

File: src/Toolbar.tsx
Comment (Irev-Dev) [reply src/Toolbar.tsx:12]: Yeah when I revert the revert
https://github.com/KittyCAD/modeling-app/pull/813
I'll do a mass clean up of this toolbar component and all of the old constraint buttons

Comment (Irev-Dev) [reply src/Toolbar.tsx:12]: 0% chance I'll forget about this mess and leave it like this haha.

File: src/Toolbar.tsx
Comment (Irev-Dev) [reply src/Toolbar.tsx:292]: Sweet, back in in 80122ca09585b4c4cb1deb9ca9e02a30a63199b9

**Comment by jessfraz** [comment]
couple of things but I dont want to delay this PR but I would hope these would be fast follow fixes and we wouldnt do a release without them

https://github.com/KittyCAD/modeling-app/assets/1445228/98283756-310b-4371-8eac-85a3dd92766a


https://github.com/KittyCAD/modeling-app/assets/1445228/a99dd334-cd9f-4a22-851a-272064e582e5

**Pull request merged**
